central/processlisteningonport/datastore/datastore_impl.go

```diff
@@ -0,0 +1,398 @@
+	}
+
+	portProcesses = normalizePLOPs(portProcesses)
+
+	// XXX: The next two calls, fetchIndicators and fetchExistingPLOPs, have to
+	// be done in a single join query fetching both ProcessIndicator and needed
+	// bits from PLOP.
+	indicatorsMap, indicatorIds, err := ds.fetchIndicators(ctx, portProcesses...)
+	if err != nil {
+		return err
+	}
+
+	existingPLOPMap, err := ds.fetchExistingPLOPs(ctx, indicatorIds, portProcesses...)
+	if err != nil {
+		return err
+	}
+
+	plopObjects := make([]*storage.ProcessListeningOnPortStorage, len(portProcesses))
+	for i, val := range portProcesses {
+		indicatorID := ""
+		var processInfo *storage.ProcessIndicatorUniqueKey
+
+		key := getProcesUniqueKey(val)
+
+		if indicator, ok := indicatorsMap[key]; ok {
+			indicatorID = indicator.GetId()
+			log.Debugf("Got indicator %s: %+v", indicatorID, indicator)
+		} else {
+			// XXX: Create a metric for this
+			log.Warnf("Found no matching indicators for %s", key)
+			processInfo = val.Process
+		}
+
+		plopKey := fmt.Sprintf("%d %d %s",
+			val.GetProtocol(),
+			val.GetPort(),
+			indicatorID,
+		)
+
+		existingPLOP, prevExists := existingPLOPMap[plopKey]
+
+		// There are three options:
+		// * We found an existing PLOP object with different close timestamp.
+		//   It has to be updated.
+		// * We found an existing PLOP object with the same close timestamp.
+		//   Nothing has to be changed (XXX: Ideally it has to be excluded from
+		//   the upsert later on).
+		// * No existing PLOP object, create a new one with whatever close
+		//   timestamp we have received and fetched indicator ID.
+		if prevExists && existingPLOP.CloseTimestamp != val.CloseTimestamp {
+			log.Debugf("Got existing PLOP: %+v", existingPLOP)
+
+			existingPLOP.CloseTimestamp = val.CloseTimestamp
+			existingPLOP.Closed = existingPLOP.CloseTimestamp != nil
+			plopObjects[i] = existingPLOP
+		}
+
+		if !prevExists {
+			if val.CloseTimestamp != nil {
+				// We try to close a not existing Endpoint, something is wrong
+				log.Warnf("Found no matching PLOP to close for %s", key)
+			}
+
+			newPLOP := &storage.ProcessListeningOnPortStorage{
+				// XXX, ResignatingFacepalm: Use regular GENERATE ALWAYS AS
+				// IDENTITY, which would require changes in store generator
+				Id:                 uuid.NewV4().String(),
+				Port:               val.Port,
+				Protocol:           val.Protocol,
+				ProcessIndicatorId: indicatorID,
+				Process:            processInfo,
+				Closed:             val.CloseTimestamp != nil,
+				CloseTimestamp:     val.CloseTimestamp,
+			}
+
+			plopObjects[i] = newPLOP
+		}
+	}
+
+	// Now save actual PLOP objects
+	err = ds.storage.UpsertMany(ctx, plopObjects)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (ds *datastoreImpl) GetProcessListeningOnPort(
+	ctx context.Context,
+	deployment string,
+) (
+	processesListeningOnPorts []*storage.ProcessListeningOnPort, err error,
+) {
+	if ok, err := plopSAC.ReadAllowed(ctx); err != nil {
```

Comment:
To be honest I'm not familiar enough with the scope access mechanism, could you show an example how to modify `ReadAllowed` call to check ClusterID and Namespace as well?

Reply:
What happened here is that there is a possibility to do SAC verification more fine grained, and allow users who have only access to certain namespaces to see PLOPs. As in example [[1]]  we would need to fetch cluster id and namespace for that. The problems would be to do this efficiently via deployment argument that we have in this API, as we only have the deployment name, which is not an indexed column in the db. All in all we agree that this could be done as a follow up, and needs to be investigated first in ROX-14417.

[1]: https://github.com/stackrox/stackrox/blob/master/central/deployment/datastore/datastore_impl.go#L129-L140